### PR TITLE
Issue #11619: The CSR subject attributes (c,st,l,o,ou,cn) will now be…

### LINKED ISSUE #11619: acmeCA-2.0: Replace RDN configuration attributes with single subject DN attribute.
The configuration attributes `country`, `state`, `locallity`, `organization` and `organizationalUnit` should be replaced by a single configuration attribute, `subjectDN`. Separate configuration attributes don't account for plurality nor any order.

### PR DESCRIPTION
**Pull request by jvanhill**
… set in a single attribute - subjectDN.

 fixes #11619

### CODE REVIEW

**Comment by jvanhill** [comment]
#build

File: dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops [hunk 1]
```diff
@@ -177,3 +177,27 @@ CWPKI2037E.useraction=Provide at least one non-null, non-empty domain in the con
 CWPKI2038W=CWPKI2038W: The ACME service was unable to revoke the certificate with serial number {0}. The error is ''{1}.''
 CWPKI2038W.explanation=The ACME service attempted to revoke the certificate, but was unable to.
 CWPKI2038W.useraction=Review the error message for details on the failure. The certificate will need to be revoked manually.
+
+CWPKI2039E=CWPKI2039E: The DN ''subjectDN'' ({0}) contains a ''cn'' relative distinguished name (RDN) value ({1}) that does not match any of the defined domains.
+CWPKI2039E.explanation= The ''cn'' RDN must match one of the defined domains.
+CWPKI2039E.useraction=Provide a ''subjectDN'' attribute value with a ''cn'' RDN value that matches one of the defined domains or does not have the ''cn'' RDN included.
+
+CWPKI2040E=CWPKI2040E: The ''cn'' relative distinguished name (RDN) was not the first RDN in the ''subjectDN'' configuration attribute.
+CWPKI2040E.explanation=If the ''cn'' RDN is defined in the ''subjectDN'' configuration attribute, then it must be the first RDN.
```
Comment (dmuelle) [inline dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops:186]: **change**
If the ''cn'' RDN is defined in the ''subjectDN'' configuration attribute, then it must be the first RDN.
**to**
If the ''cn'' RDN is defined, it must be the first RDN  in the ''subjectDN'' configuration attribute.

File: dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops [hunk 2]
```diff
@@ -177,3 +177,27 @@ CWPKI2037E.useraction=Provide at least one non-null, non-empty domain in the con
...
+CWPKI2040E.useraction=Provide a ''subjectDN'' attribute value that either defines the ''cn'' attribute as the first RDN, or does not define the ''cn'' attribute as an RDN. If it is not defined, the first defined domain will be used.
```
Comment (dmuelle) [inline dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops:187]: **change**
Provide a ''subjectDN'' attribute value that either defines the ''cn'' attribute as the first RDN, or does not define the ''cn'' attribute as an RDN. If it is not defined, the first defined domain will be used.

**To**
Provide a ''subjectDN'' attribute value that either defines the ''cn'' attribute as the first RDN, or does not the ''cn'' attribute. If the ''cn'' attribute is not defined, the first defined domain is used as the subject distinguished name.

File: dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops [hunk 3]
```diff
@@ -177,3 +177,27 @@ CWPKI2037E.useraction=Provide at least one non-null, non-empty domain in the con
...
+
+CWPKI2041E=CWPKI2041E: The relative distinguished name (RDN) type ''{0}'' in the ''subjectDN'' attribute is not supported. The following RDN types ae supported: ''cn'', ''o'', ''ou'', ''c'', ''st'', ''l''
```
Comment (dmuelle) [inline dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops:189]: **change**
The relative distinguished name (RDN) type ''{0}'' in the ''subjectDN'' attribute is not supported
**to**
The {0} relative distinguished name (RDN) type in the ''subjectDN'' attribute is not supported

File: dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops [hunk 4]
```diff
@@ -177,3 +177,27 @@ CWPKI2037E.useraction=Provide at least one non-null, non-empty domain in the con
...
+CWPKI2041E.explanation=The ''subjectDN'' attribute value contains an RDN that is not supported.
+CWPKI2041E.useraction=Provide a ''subjectDN'' attribute value that does not contain unsupported RDN types.
```
Comment (dmuelle) [inline dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops:191]: **change**
Provide a ''subjectDN'' attribute value that does not contain unsupported RDN types.
**To**
Provide a ''subjectDN'' attribute value that does not contain unsupported RDN types. The following RDN types ae supported: ''cn'', ''o'', ''ou'', ''c'', ''st'',  ''l''.

File: dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops [hunk 5]
```diff
@@ -177,3 +177,27 @@ CWPKI2037E.useraction=Provide at least one non-null, non-empty domain in the con
...
+
+CWPKI2042E=CWPKI2042E: The ''subjectDN'' attribute value ''{0}'' is not a valid distinguished name. The error is ''{1}''.
```
Comment (dmuelle) [inline dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops:193]: **change**
The ''subjectDN'' attribute value ''{0}'' is not a valid distinguished name.
**to**
The {0} ''subjectDN'' attribute value is not a valid distinguished name.

File: dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops [hunk 6]
```diff
@@ -177,3 +177,27 @@ CWPKI2037E.useraction=Provide at least one non-null, non-empty domain in the con
...
+CWPKI2042E.explanation= The ''subjectDN'' attribute must be a valid distinguished name.
+CWPKI2042E.useraction= Review the error message for details on the failure.
+
+CWPKI2043E=CWPKI2043E: The value ''{0}'' is not a valid RDN. The error is ''{1}''.
```
Comment (dmuelle) [inline dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops:197]: **change**
 The value ''{0}'' is not a valid RDN. The error is ''{1}''.
**to**
 The {0} value is not a valid RDN. The error is {1}.

File: dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops [hunk 7]
```diff
@@ -177,3 +177,27 @@ CWPKI2037E.useraction=Provide at least one non-null, non-empty domain in the con
...
+CWPKI2043E.explanation= The ACME service was not able to create an RDN from the specified value.
+CWPKI2043E.useraction= Review the error message for details on the failure.
+
+CWPKI2044E=CWPKI2044E: The certificate is not an X.509 certificate. The certificate type is ''{0}''.
```
Comment (dmuelle) [inline dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops:201]: **change**
The certificate type is ''{0}''.
**to**
The certificate type is {0}.

File: dev/com.ibm.ws.security.acme/resources/OSGI-INF/l10n/metatype.properties [hunk 8]
```diff
@@ -25,23 +25,8 @@ domain.desc=A domain name to request a certificate for.
...
-organization=Organization
-organization.desc=Organization name to use for the certificate.
-
-organizationalUnit=Organizational unit
-organizationalUnit.desc=Organizational unit name to use for the certificate.
-
-acceptTermsOfService=Accept terms of service
-acceptTermsOfService.desc=Indicates whether the user accepts the terms of service for the ACME CA server. If true, the terms of service are automatically accepted. If false, and the ACME CA server issues terms of service, the request for a certificate fails.  
+subjectDN=Subject distinguished name
+subjectDN.desc=Subject distinguished name to use for the certificate. The DN may include the following relative distinguished name types: ''cn'', ''c'', ''st'', ''l'', ''o'' and ''ou''. The ''cn'' relative distinguished name type, if defined, is required to be first and must be one of the defined domains. If not defined, the first defined domain will be used.
```
[the remark this answers is not in the document]
Comment (jvanhill) [reply dev/com.ibm.ws.security.acme/resources/OSGI-INF/l10n/metatype.properties:29]: ```subjectDN.desc=Subject distinguished name (DN) to use for the certificate. The DN can include the following relative distinguished name (RDN) types: cn, c, st, l, o and ou. If the cn RDN type is defined, it must be one of the domains defined by the domain configuration element and it must be the first RDN in the DN. If the cn RDN type is not defined, the first domain defined by the domain configuration element is used as the cn RDN value.```

File: dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops [hunk 9]
```diff
@@ -177,3 +177,27 @@ CWPKI2037E.useraction=Provide at least one non-null, non-empty domain in the con
 CWPKI2038W=CWPKI2038W: The ACME service was unable to revoke the certificate with serial number {0}. The error is ''{1}.''
 CWPKI2038W.explanation=The ACME service attempted to revoke the certificate, but was unable to.
 CWPKI2038W.useraction=Review the error message for details on the failure. The certificate will need to be revoked manually.
+
+CWPKI2039E=CWPKI2039E: The DN ''subjectDN'' ({0}) contains a ''cn'' relative distinguished name (RDN) value ({1}) that does not match any of the defined domains.
```
[the remark this answers is not in the document]
Comment (jvanhill) [reply dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops:181]: ```CWPKI2039E: The ''{0}'' distinguished name (DN) defined by subjectDN contains a ''{1}'' cn relative distinguished name (RDN) value that does not match any of the defined domains.```

File: dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops [hunk 10]
```diff
@@ -50,7 +50,7 @@ CWPKI2006I=CWPKI2006I: The ACME certificate authority at the {0} URI provided th
 CWPKI2006I.explanation=The certificate authority provides terms of service.
 CWPKI2006I.useraction= Review the provided terms of service.
 
-CWPKI2007I=CWPKI2007I: The ACME service installed a certificate that is signed by the ACME certificate authority at the {0} URI. The expiration date is {1}.
+CWPKI2007I=CWPKI2007I: The ACME service installed a certificate with serial number {0} that is signed by the ACME certificate authority at the {1} URI. The expiration date is {2}.
```
[the remark this answers is not in the document]
Comment (jvanhill) [reply dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops:53]: Agreed.

File: dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops [hunk 11]
```diff
@@ -177,3 +177,27 @@ CWPKI2037E.useraction=Provide at least one non-null, non-empty domain in the con
 CWPKI2038W=CWPKI2038W: The ACME service was unable to revoke the certificate with serial number {0}. The error is ''{1}.''
 CWPKI2038W.explanation=The ACME service attempted to revoke the certificate, but was unable to.
 CWPKI2038W.useraction=Review the error message for details on the failure. The certificate will need to be revoked manually.
+
+CWPKI2039E=CWPKI2039E: The DN ''subjectDN'' ({0}) contains a ''cn'' relative distinguished name (RDN) value ({1}) that does not match any of the defined domains.
+CWPKI2039E.explanation= The ''cn'' RDN must match one of the defined domains.
+CWPKI2039E.useraction=Provide a ''subjectDN'' attribute value with a ''cn'' RDN value that matches one of the defined domains or does not have the ''cn'' RDN included.
```
[the remark this answers is not in the document]
Comment (jvanhill) [reply dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops:183]: Agreed.

File: dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops (hunk 1 above)
Comment (jvanhill) [reply dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops:186]: Agreed.

File: dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops (hunk 2 above)
Comment (jvanhill) [reply dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops:187]: Slight change at the end for clarity.

```CWPKI2040E.useraction=Provide a subjectDN attribute value that either defines the cn attribute as the first RDN, or does not contain the cn attribute. If the cn attribute is not defined, the first defined domain is used as the cn RDN value.```

File: dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops (hunk 3 above)
Comment (jvanhill) [reply dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops:189]: ```CWPKI2041E=CWPKI2041E: The {0} relative distinguished name (RDN) type in the ''subjectDN'' attribute is not supported. The following RDN types are supported: cn, o, ou, c, st, l```

File: dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops (hunk 4 above)
Comment (jvanhill) [reply dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops:191]: ```CWPKI2041E.useraction=Provide a subjectDN attribute value that does not contain unsupported RDN types. The following RDN types are supported: cn, o, ou, c, st, l.```

File: dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops (hunk 5 above)
Comment (jvanhill) [reply dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops:193]: ```CWPKI2042E=CWPKI2042E: The ''{0}'' subjectDN attribute value is not a valid distinguished name. The error is ''{1}''.```

File: dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops (hunk 6 above)
Comment (jvanhill) [reply dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops:197]: ```CWPKI2043E=CWPKI2043E: The ''{0}'' relative distinguished name (RDN) is not valid. The error is ''{1}''.```

File: dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops (hunk 7 above)
Comment (jvanhill) [reply dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops:201]: Agreed.

File: dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops [hunk 12]
```diff
@@ -177,3 +177,27 @@ CWPKI2037E.useraction=Provide at least one non-null, non-empty domain in the con
 CWPKI2038W=CWPKI2038W: The ACME service was unable to revoke the certificate with serial number {0}. The error is ''{1}.''
 CWPKI2038W.explanation=The ACME service attempted to revoke the certificate, but was unable to.
 CWPKI2038W.useraction=Review the error message for details on the failure. The certificate will need to be revoked manually.
+
+CWPKI2039E=CWPKI2039E: The ''{0}'' distinguished name (DN) defined by subjectDN contains a ''{1}'' cn relative distinguished name (RDN) value that does not match any of the defined domains.
+CWPKI2039E.explanation=The cn RDN must match one of the defined domains.
+CWPKI2039E.useraction=Provide a subjectDN attribute value with a cn RDN value that matches one of the defined domains or does not have the cn RDN included.
```
Comment (dmuelle) [inline dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops:183]: **change to**
Provide either a subjectDN attribute value with a cn RDN value that matches one of the defined domains or a subjectDN attribute value that does not have the cn RDN included.

File: dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops [hunk 13]
```diff
@@ -177,3 +177,27 @@ CWPKI2037E.useraction=Provide at least one non-null, non-empty domain in the con
...
+CWPKI2041E.explanation=The subjectDN attribute value contains an RDN that is not supported.
+CWPKI2041E.useraction=Provide a subjectDN attribute value that does not contain unsupported RDN types. The following RDN types are supported: cn, o, ou, c, st, l.
+
+CWPKI2042E=CWPKI2042E: The ''{0}'' subjectDN attribute value is not a valid distinguished name. The error is ''{1}''.
+CWPKI2042E.explanation=The subjectDN attribute must be a valid distinguished name.
+CWPKI2042E.useraction=Provide a valid distinguished name as the subjectDN attribute value.
+
+CWPKI2043E=CWPKI2043E: The ''{0}'' value is not a valid relative distinguished name (RDN). The error is ''{1}''.
+CWPKI2043E.explanation=The ACME service was not able to create an RDN from the specified value.
+CWPKI2043E.useraction=Review the error message for details on the failure.
```
Comment (dmuelle) [inline dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops:199]: Can you give any info about what constitutes a valid RDN? The user action should offer more than just "see error message", which is a given.

File: dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops [hunk 14]
```diff
@@ -177,3 +177,27 @@ CWPKI2037E.useraction=Provide at least one non-null, non-empty domain in the con
...
+
+CWPKI2044E=CWPKI2044E: The certificate is not an X.509 certificate. The certificate type is {0}.
+CWPKI2044E.explanation=The ACME service expects all certificates in the certificate chain to be X.509 certificates.
+CWPKI2044E.useraction=Verify that the certificate chain is comprised of X.509 certificates and try again.
```
Comment (dmuelle) [inline dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops:203]: **change**
Verify that the certificate chain is comprised of X.509 certificates and try again.
**to**
Ensure that all the certificates in the certificate chain are X.509 certificates and try again.

**Comment by jvanhill** [comment]
#libby

File: dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops (hunk 12 above)
Comment (jvanhill) [reply dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops:183]: agreed

File: dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops (hunk 14 above)
Comment (jvanhill) [reply dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops:203]: agreed

File: dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops (hunk 13 above)
Comment (jvanhill) [reply dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops]: ```CWPKI2043E.useraction=Ensure the the domain included in the value is a valid RDN value.```

**Comment by jvanhill** [comment]
#libby

File: dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops [hunk 15]
```diff
@@ -177,3 +177,27 @@ CWPKI2037E.useraction=Provide at least one non-null, non-empty domain in the con
 CWPKI2038W=CWPKI2038W: The ACME service was unable to revoke the certificate with serial number {0}. The error is ''{1}.''
 CWPKI2038W.explanation=The ACME service attempted to revoke the certificate, but was unable to.
 CWPKI2038W.useraction=Review the error message for details on the failure. The certificate will need to be revoked manually.
+
+CWPKI2039E=CWPKI2039E: The ''{0}'' distinguished name (DN) defined by subjectDN contains a ''{1}'' cn relative distinguished name (RDN) value that does not match any of the defined domains.
+CWPKI2039E.explanation=The cn RDN must match one of the defined domains.
```
Comment (dmuelle) [inline dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops:182]: If the cn RDN value is included in the DN, it must match one of the defined domains.

Comment (jvanhill) [reply dev/com.ibm.ws.security.acme/resources/com/ibm/ws/security/acme/resources/AcmeMessages.nlsprops]: agreed

**Comment by jvanhill** [comment]
#libby

**Comment by jvanhill** [comment]
#build

**Comment by jvanhill** [comment]
Test failures in the last build were all GUI tests that are not related to ACME or the changes I made.

**Pull request merged**